eng/common/pipelines/templates/steps/sparse-checkout.yml

```diff
@@ -16,23 +16,54 @@ steps:
-        git sparse-checkout init
-        $paths = ('${{ convertToJson(parameters.Paths) }}' | ConvertFrom-Json) -Join ' '
-        Invoke-Expression -Command "git sparse-checkout set eng $paths"
-        Write-Host "Set sparse checkout paths to:"
-        Get-Content .git/info/sparse-checkout
-      displayName: Init sparse checkout ${{ repo.Name }}
-      workingDirectory: ${{ coalesce(repo.WorkingDirectory, format('{0}/{1}', '$(System.DefaultWorkingDirectory)', repo.Name)) }}
-
-    - pwsh: |
-        git checkout ${{ repo.Commitish }}  # this will use the default branch if repo.Commitish is empty
-      displayName: Sparse checkout at ${{ coalesce(repo.Commitish, 'default branch') }}
-      workingDirectory: ${{ coalesce(repo.WorkingDirectory, format('{0}/{1}', '$(System.DefaultWorkingDirectory)', repo.Name)) }}
+  - task: PowerShell@2
+    displayName: 'Sparse checkout repositories'
+    inputs:
+      targetType: inline
+      # Define this inline, because of the chicken/egg problem with loading a script when nothing
+      # has been checked out yet.
+      script: |
+        function SparseCheckout([Array]$paths, [Array]$Repositories)
+        {
+            $paths = $paths -Join ' '
+
+            foreach ($repo in $Repositories) {
+                $dir = $repo.WorkingDirectory
+                if (!$dir) {
+                  $dir = "./$($repo.Name)"
+                }
+                New-Item $dir -ItemType Directory -Force
+                Push-Location $dir
+
+                if (Test-Path .git/info/sparse-checkout) {
+                  $hasInitialized = $true
+                  Write-Host "Repository $($repo.Name) has already been initialized. Skipping this step."
+                } else {
+                  Write-Host "Repository $($repo.Name) is being initialized."
+                  git clone --no-checkout --filter=tree:0 git://github.com/$($repo.Name) .
+                  git sparse-checkout init
+                  git sparse-checkout set eng
+                }
+
+                Invoke-Expression -Command "git sparse-checkout add $paths"
+
+                Write-Host "Set sparse checkout paths to:"
+                Get-Content .git/info/sparse-checkout
+
+                # sparse-checkout commands after initial checkout will auto-checkout again
+                if (!$hasInitialized) {
+                  git checkout $($repo.Commitish)  # this will use the default branch if repo.Commitish is empty
+                }
+
+                Pop-Location
+            }
+        }
+
+        # Paths may be sourced as a yaml object literal OR a dynamically generated variable json string.
+        # If the latter, convertToJson will wrap the 'string' in quotes, so remove them.
+        $paths = '${{ convertToJson(parameters.Paths) }}'.Trim('"') | ConvertFrom-Json
+        # Escape windows backslash paths
+        $repositories = '${{ convertToJson(parameters.Repositories) }}' -replace '\\', '\\' | ConvertFrom-Json
```

Comment:
Is that replace regex correct? Should we simply replace all the slash with `/` as that should work on both windows and linux and doesn't require escaping.